app/controllers/api/graphql_controller.rb

```diff
@@ -28,6 +28,20 @@ def api_request?
       true
     end
 
+    def authenticate
+      if bruteforce_attempt?
+        log_bruteforce
+        render_error('Bruteforce attempt.', status: :unauthorized)
+        return false
+      end
+
+      authenticated = super
+
+      count_login_failure if available_sso.present? && !authenticated
+
+      authenticated
```

Comment:
this would be `nil` if the authentication failed, i'm guessing it would still stop the action because it's falsy but perhaps it's better to be more explicit about it (like we do in the api base controller)?

Reply:
hmm, interesting. Turns out that according to [the manual](https://guides.rubyonrails.org/action_controller_overview.html#filters) only redirect or renders stop the action, and doesn't looks like [there is any use for the return values](https://github.com/rails/rails/blob/b9ca94caea2ca6a6cc09abaffaad67b447134079/actionpack/lib/abstract_controller/callbacks.rb#L34). Even after almost 5 years of writing Rails I keep learning new stuff.

Should we render an error response if the authentication failed? or do we and i missed it?

Reply:
I guess we should add better (and graphql compliant) error handling, that's still an open todo item. So I would do it properly in a separate PR.